notification-hubs-sdk/src/main/java/com/microsoft/windowsazure/messaging/notificationhubs/TemplateVisitor.java

```diff
@@ -0,0 +1,217 @@
+package com.microsoft.windowsazure.messaging.notificationhubs;
+
+import android.content.Context;
+import android.content.SharedPreferences;
+
+import com.microsoft.windowsazure.messaging.R;
+
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Collects a set of distinct templates, in order to apply them to {@link Installation}s as they are
+ * created.
+ */
+public class TemplateVisitor implements InstallationVisitor {
+
+    private static final String PREFERENCE_KEY = "templates";
+    private SharedPreferences mPreferences;
+
+    /**
+     * Creates an empty TemplateVisitor.
+     */
+    public TemplateVisitor() {
+
+    }
+
+    /**
+     * Creates a TemplateVisitor with a pre-populated set of templates to apply.
+     * @param context application context
+     * @param templates The initial set of templates that should be applied to future {@link Installation}s.
+     */
+    public TemplateVisitor(Context context, Map<String, InstallationTemplate> templates) {
+        this();
+        setPreferences(context);
+        addTemplates(templates);
+    }
+
+    @Override
+    public void visitInstallation(Installation subject) {
+        subject.addTemplates(getSharedPreferenceTemplates());
+    }
+
+    /**
+     * Get a set of templates.
+     *
+     * @return A map of templates.
+     */
+    private Map<String, InstallationTemplate> getSharedPreferenceTemplates() {
+        Set<String> preferencesSet = mPreferences.getStringSet(PREFERENCE_KEY, new HashSet<String>());
+        if (preferencesSet == null) {
+            return new HashMap<>();
+        }
+        return deserializeInstallationTemplateToJson(preferencesSet);
+    }
+
+    /**
+     * Save templates to shared preference.
+     */
+    void setPreferences(Context context) {
+        mPreferences = context.getSharedPreferences(String.valueOf(R.string.installation_enrichment_file_key), Context.MODE_PRIVATE);
+    }
```

Comment:
Recently I've moved away from having a `setPreferences` method. Let's just move this into the currently empty constructor and have it take Context.